kotlinx-coroutines-core/common/src/internal/Scopes.kt

```diff
@@ -19,18 +20,16 @@ internal open class ScopeCoroutine<in T>(
     final override fun getStackTraceElement(): StackTraceElement? = null
     final override val isScopedCoroutine: Boolean get() = true
 
-    override val defaultResumeMode: Int get() = MODE_DIRECT
-
     internal val parent: Job? get() = parentContext[Job]
 
-    @Suppress("UNCHECKED_CAST")
-    override fun afterCompletionInternal(state: Any?, mode: Int) {
-        if (state is CompletedExceptionally) {
-            val exception = if (mode == MODE_IGNORE) state.cause else recoverStackTrace(state.cause, uCont)
-            uCont.resumeUninterceptedWithExceptionMode(exception, mode)
-        } else {
-            uCont.resumeUninterceptedMode(state as T, mode)
-        }
+    override fun afterCompletion(state: Any?) {
+        // Resume in a cancellable way by default when resuming from another context
+        uCont.intercepted().resumeCancellableWith(recoverResult(state, uCont))
```

Comment:
Yes. We'll need to rework this logic. Instead of recovering in `DispatchedTask.run`, we should recover in all places where where we covered coroutine state or other exception into `kotlin.Result` and call some `resumeXxxWith(....)` method. This should also close all the places where we don't currently recover exceptions at all.